### PR TITLE
fix: [CDS-71328]: make K8s api client timeouts configurable

### PR DESCRIPTION
**Pull request by abhinavsingh1196**
## Describe your changes
**TL;DR - A bug in JDK11 does not pass interrupts when doing native socketRead0 call. This is resulting in a thread leak. We need to set timeout in K8s API client to fix this.**

A customer reported an unresponsive delegate where their k8s executions were getting stuck at Prepare step while fetching the existing pods from a cluster (Detailed analysis here: https://harness.atlassian.net/browse/CDS-71328?focusedCommentId=512952).
Most of Harness' K8s API calls do not specify fixed timeouts in the okhttp client. Setting this timeout would fix this thread leak. 

**Changelog**:
- Make K8s read and connect timeouts configurable via env variables
- Use timeout based client in all K8s API calls, except K8s watch calls
- Removed some unused code

Tests:
- Reproduced customer scenario by simulating network faults, thread is now being released as expected (see `default-task-2` thread below).
Screenshot 2023-06-08 at 7 29 31 PM
- Ran a long running (4 min) steady state check based k8s deployment.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48994)

### CODE REVIEW

**Comment by abhinavsingh1196** [comment]
trigger uts

**Review by alebuh — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48994)* status: 7 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): k8s-team. 

___
*[`960-api-services/src/main/java/io/harness/k8s/apiclient/ApiClientFactoryImpl.java` line 123 at r2](https://reviewable.io/reviews/harness/harness-core/48994#-NXTp63B8trXQ7tHoKRI:-NXTp63C5WcCd3mHZCRa:batt806) ([raw file](https://github.com/harness/harness-core/blob/fce9462b7678dcd58a2259d10a04999f4580272c/960-api-services/src/main/java/io/harness/k8s/apiclient/ApiClientFactoryImpl.java#L123)):*
> ```Java
>     long readTimeout = K8sApiClientHelper.getOrDefaultTimeout(READ_TIMEOUT_ENV_VAR, DEFAULT_READ_TIMEOUT_SECONDS);
> 
>     // don't timeout on client-side
> ```

nit: I think this comment can be removed now to avoid confusion as there are some cases when we are doing timeout on the client side
___
*[`960-api-services/src/main/java/io/harness/k8s/apiclient/K8sApiClientHelper.java` line 66 at r2](https://reviewable.io/reviews/harness/harness-core/48994#-NXTqapH2bML-tKhO25D:-NXTqapH2bML-tKhO25E:b9ers90) ([raw file](https://github.com/harness/harness-core/blob/fce9462b7678dcd58a2259d10a04999f4580272c/960-api-services/src/main/java/io/harness/k8s/apiclient/K8sApiClientHelper.java#L66)):*
> ```Java
>   }
> 
>   public long getOrDefaultTimeout(String environmentVariable, long defaultTimeout) {
> ```

We are providing defaultTimeout in this method but we are not doing anything with it, so why pass it at all?
My suggestion would be that this method return Optional<Long> and that it returns Optional.empty() if it's not able to resolve the env variable. And when this method is called we can use orElse(DEFAULT_TIMEOUT)

**Comment by abhinavsingh1196** [comment]
trigger uts

**Review by abhinavsingh1196 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48994)* status: 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): k8s-team. 

___
*[`960-api-services/src/main/java/io/harness/k8s/apiclient/ApiClientFactoryImpl.java` line 123 at r2](https://reviewable.io/reviews/harness/harness-core/48994#-NXTp63B8trXQ7tHoKRI:-NXU0ZHtDldA7bq362p6:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/fce9462b7678dcd58a2259d10a04999f4580272c/960-api-services/src/main/java/io/harness/k8s/apiclient/ApiClientFactoryImpl.java#L123)):*
<details><summary><i>Previously, alebuh (Aleksa Buha) wrote…</i></summary><blockquote>

nit: I think this comment can be removed now to avoid confusion as there are some cases when we are doing timeout on the client side
</blockquote></details>

Done
___
*[`960-api-services/src/main/java/io/harness/k8s/apiclient/K8sApiClientHelper.java` line 66 at r2](https://reviewable.io/reviews/harness/harness-core/48994#-NXTqapH2bML-tKhO25D:-NXU0dyJ5kQgE8CaAAHS:b78ay8s) ([raw file](https://github.com/harness/harness-core/blob/fce9462b7678dcd58a2259d10a04999f4580272c/960-api-services/src/main/java/io/harness/k8s/apiclient/K8sApiClientHelper.java#L66)):*
<details><summary><i>Previously, alebuh (Aleksa Buha) wrote…</i></summary><blockquote>

We are providing defaultTimeout in this method but we are not doing anything with it, so why pass it at all?
My suggestion would be that this method return Optional<Long> and that it returns Optional.empty() if it's not able to resolve the env variable. And when this method is called we can use orElse(DEFAULT_TIMEOUT)
</blockquote></details>

Nice suggestion, thanks! I have removed this param from the method now.

**Review by alebuh — COMMENTED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48994)* status: 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): k8s-team.

**Review by alebuh — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48994)* status: 7 file(s) to review. 1 of 2 approvals obtained. Waiting on team(s): k8s-team.

**Review by abosii-harness — APPROVED** [review]

**Comment by abhinavsingh1196** [comment]
trigger review

**Comment by abhinavsingh1196** [comment]
trigger messagecheck
